lib/middleware.ts

```diff
@@ -0,0 +1,120 @@
+'use strict'
+
+/**
+ * Module dependencies.
+ * @private
+ */
+
+var kayvee = require('../lib/kayvee')
+var morgan = require('morgan')
+var _ = require('underscore')
+
+/**
+ * request path
+ */
+
+var getBaseUrl = function(req) {
+  var url = req.originalUrl || req.url
+  var parsed = require('url').parse(url, true)
+  return parsed.pathname
+}
+
+/**
+ * request query params
+ */
+
+var getQueryParams = function(req) {
+  var url = req.originalUrl || req.url
+  var parsed = require('url').parse(url, true)
+  return parsed.search
+}
+
+/**
+ * response size
+ */
+
+var getResponseSize = function(res) {
+  var headers = res.headers || res._headers
+  if (headers && headers['content-length']) {
+    return Number(headers['content-length'])
+  } else if (res.data) {
+    return res.data.length
+  }
+}
+
+/**
+ * response time in nanoseconds
+ */
+
+var getResponseTimeNs = function(req, res) {
+  if (!req._startAt || !res._startAt) {
+    // missing request and/or response start time
+    return
+  }
+
+  // calculate diff
+  var ns = (res._startAt[0] - req._startAt[0]) * 1e9
```

Comment:
This looks like it's per-second resolution, and we're converting it into nanoseconds? Is this going to be granular enough?